### PR TITLE
add mask_src_tokens parameter to preference finetuning

### PR DESCRIPTION
**Pull request by asn1814**
**What does this PR do? Please describe:**
Allows src tokens to be used in the loss calculation for preference finetuning recipes, which may be beneficial when dropout is used as shown by Ilia. 

Fixes #{issue number}

**Does your PR introduce any breaking changes? If yes, please list them:**
None

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/datasets/preference.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ def create_reader(
         max_num_batches: int | None = None,
         num_accumulate: int = 1,
         num_prefetch: int = 1,
+        mask_src_tokens: bool = True,
```
Comment (cbalioglu) [inline src/fairseq2/datasets/preference.py:66]: Nit: in the rest of the code base we use `source` instead of `src` in variable names. So `mask_source_tokens` would be more consistent.

Comment (cbalioglu) [reply src/fairseq2/datasets/preference.py:66]: Also, can you please a brief docstr for this parameter?

File: src/fairseq2/datasets/preference.py [hunk 2]
```diff
@@ -215,7 +217,9 @@ def cat_source_and_target(example: dict[str, Any]) -> dict[str, Any]:
             indices_chosen = torch.cat([prompt_indices, target_indices_chosen])
             indices_rejected = torch.cat([prompt_indices, target_indices_rejected])
 
-            prompt_len = len(prompt_indices)
+            prompt_len = 0
+            if mask_src_tokens:
```
Comment (cbalioglu) [inline src/fairseq2/datasets/preference.py:221]: I think lines 224 and 225 can be slightly optimized by skipping `>=` filtering all together if `mask_src_tokens` is true.

**Review by cbalioglu — COMMENTED** [review]
Left three nit comments. Once you address them, it is good to go.

File: src/fairseq2/datasets/preference.py (hunk 1 above)
Comment (asn1814) [reply src/fairseq2/datasets/preference.py:66]: thanks for docstr catch!

**Review by cbalioglu — APPROVED** [review]
LGTM!
